src/ScheduleFunctions.cpp

```diff
@@ -354,9 +354,21 @@ Stmt build_provide_loop_nest(const map<string, Function> &env,
     // Make the (multi-dimensional multi-valued) store node.
     Stmt body = Provide::make(func.name(), values, site);
     if (def.schedule().atomic()) {  // Add atomic node.
-        // If required, we will allocate a mutex buffer called func.name() + ".mutex"
-        // The buffer is added in the AddAtomicMutex pass.
-        body = Atomic::make(func.name(), func.name() + ".mutex", body);
+        bool any_unordered_parallel = false;
+        for (auto d : def.schedule().dims()) {
+            any_unordered_parallel |= is_unordered_parallel(d.for_type);
+        }
+        if (any_unordered_parallel) {
+            // If required, we will allocate a mutex buffer called func.name() + ".mutex"
+            // The buffer is added in the AddAtomicMutex pass.
+            body = Atomic::make(func.name(), func.name() + ".mutex", body);
+        } else {
+            // No mutex is required if there is no parallelism, and it
+            // doesn't work if all parallelism is synchronous
```

Comment:
does "doesn't work at all" mean "will fail to compile" or "will give bad results unexpectedly"? if the latter, we should inject checks

Reply:
I'll clarify the comment. It's more that it wouldn't work so we don't do it rather than it doesn't work. 

This PR would just scalarize such code during vectorization anyway.